File: language_comparative_table/nlp100_ch1.js
var DATA = require('./data.js');

function nlp00(){
    var s = "stressed";
    return s.split("").reverse().join("");
    // 文字列長が10^4程度であれば以下のコードより速い
    // 文字列長が10^5を超える場合は以下のコードの方が速い
    // var rv = "";
    // var n = s.length;
    // for(var i=0; i<n; ++i){
    //     rv += s[n-1-i];
    // }
    // return rv;
}

function nlp01(){
    var s = "パタトクカシーー";
    var res = "";
    for (var i=0; i<s.length; i+=2){
        res += s[i];
    }
    return res;
}

function nlp02(){
    var s = ["パトカー", "タクシー"];
    var res = "";
    for (var i=0; i<s[0].length; ++i){
        for(var j=0; j<s.length; ++j){
            res += s[j][i];
        }
    }
    return res;
}
function nlp03(){
}
function nlp04(){
}
function nlp05(){
    return "";
}
function nlp06(){
    return "";
}
function nlp07(){
    return "";
}
function nlp08(){
    return "";
}
function nlp09(){
    return "";
}

function measureElapsedTime(func){
    var start = new Date().getTime();

    var res = func();

    var end = new Date().getTime();
    console.log(func.name);
    console.log("Elapsed Time: " + (end - start) + "[ms]");
    console.log(res);
}

function main(){
    measureElapsedTime(nlp00);
    measureElapsedTime(nlp01);
    measureElapsedTime(nlp02);
    measureElapsedTime(nlp03);
    measureElapsedTime(nlp04);
    measureElapsedTime(nlp05);
    measureElapsedTime(nlp06);
    measureElapsedTime(nlp07);
    measureElapsedTime(nlp08);
    measureElapsedTime(nlp09);
}
main();

File: language_comparative_table/nlp100_ch1.py
def nlp_00():
    return 'stressed'[::-1]


def nlp_01():
    return 'パタトクカシーー'[::2]


def nlp_02():
    a = 'パトカー'
    b = 'タクシー'
    return ''.join([a[i] + b[i] for i in range(len(a))])


def nlp_03():
    s = 'Now I need a drink, alcoholic of course, after the heavy lectures involving quantum mechanics.'
    return [len(word) for word in s.replace(',', '').replace('.', '').split()]


def nlp_04():
    s = 'Hi He Lied Because Boron Could Not Oxidize Fluorine. New Nations Might Also Sign Peace Security Clause. Arthur King Can.'
    table = {}
    for i, word in enumerate(s.replace(',', '').replace('.', '').split()):
        if i + 1 in [1, 5, 6, 7, 8, 9, 15, 16, 19]:
            table.update({word[:1]: i + 1})
        else:
            table.update({word[:2]: i + 1})
    return table


def get_ngram(sequence, n=2):
    ngrams = []
    for i in range(len(sequence) - n + 1):
        ngrams.append(sequence[i:i + n])
    return ngrams


def nlp_05():
    s = 'I am an NLPer'
    return get_ngram(s), get_ngram(s.split())


def nlp_06():
    s = ['paraparaparadise', 'paragraph']
    X = set(get_ngram(s[0]))
    Y = set(get_ngram(s[1]))

    return X | Y, X & Y, X - Y, Y - X, 'se' in X, 'se' in Y


def nlp_07():
    def nlp_07_template(x, y, z):
        return str(x) + '時の' + str(y) + 'は' + str(z)
    return nlp_07_template(12, '気温', 22.4)


def nlp_08():
    def cipher(s):
        import string
        res = ''
        for c in s:
            if c in string.ascii_lowercase:
                res += chr(219 - ord(c))
            else:
                res += c
        return res
    return cipher('日本 is very Japan.'), cipher(cipher('日本 is very Japan.'))


def nlp_09():
    s = "I couldn't believe that I could actually understand what I was reading : the phenomenal power of the human mind ."

    def typoglycemia(s):
        import random
        res = []
        for word in s.split():
            if len(word) <= 4:
                res.append(word)
            else:
                mid = [c for c in word[1:-1]]
                random.shuffle(mid)
                res.append(word[0] + ''.join(mid) + word[-1])
        return ' '.join(res)
    return typoglycemia(s)


if __name__ == '__main__':
    print(nlp_00())
    print(nlp_01())
    print(nlp_02())
    print(nlp_03())
    print(nlp_04())
    print(nlp_05())
    print(nlp_06())
    print(nlp_07())
    print(nlp_08())
    print(nlp_09())
